src/github.com/couchbase/sync_gateway/db/change_cache.go

```diff
@@ -48,25 +56,67 @@ type LogEntries []*LogEntry
 // A priority-queue of LogEntries, kept ordered by increasing sequence #.
 type LogPriorityQueue []*LogEntry
 
+// An ordered queue of supporting Remove
+type SkippedSequenceQueue []*SkippedSequence
+
+type SkippedSequence struct {
+	seq       uint64
+	timeAdded time.Time
+}
+
+type CacheOptions struct {
+	CachePendingSeqMaxWait time.Duration // Max wait for pending sequence before skipping
+	CachePendingSeqMaxNum  int           // Max number of pending sequences before skipping
+	CacheSkippedSeqMaxWait time.Duration // Max wait for skipped sequence before abandoning
+}
+
 //////// HOUSEKEEPING:
 
 // Initializes a new changeCache.
 // lastSequence is the last known database sequence assigned.
 // onChange is an optional function that will be called to notify of channel changes.
-func (c *changeCache) Init(context *DatabaseContext, lastSequence uint64, onChange func(base.Set)) {
+func (c *changeCache) Init(context *DatabaseContext, lastSequence uint64, onChange func(base.Set), options CacheOptions) {
 	c.context = context
 	c.initialSequence = lastSequence
 	c.nextSequence = lastSequence + 1
 	c.onChange = onChange
 	c.channelCaches = make(map[string]*channelCache, 10)
 	c.receivedSeqs = make(map[uint64]struct{})
+
+	// init cache options
+	c.options = CacheOptions{
+		CachePendingSeqMaxWait: DefaultMaxChannelLogPendingWaitTime,
+		CachePendingSeqMaxNum:  DefaultMaxChannelLogPendingCount,
+		CacheSkippedSeqMaxWait: DefaultMaxChannelLogSkippedWaitTime,
+	}
+
+	if options.CachePendingSeqMaxNum > 0 {
+		c.options.CachePendingSeqMaxNum = options.CachePendingSeqMaxNum
+	}
+
+	if options.CachePendingSeqMaxWait > 0 {
+		c.options.CachePendingSeqMaxWait = options.CachePendingSeqMaxWait
+	}
+
+	if options.CacheSkippedSeqMaxWait > 0 {
+		c.options.CacheSkippedSeqMaxWait = options.CacheSkippedSeqMaxWait
+	}
+
+	base.LogTo("Cache", "Initializing changes cache with options %v", c.options)
```

Comment:
if `%+v` is used instead of `%v`, it will print out name:value pairs of struct rather than unnamed values, which should be easier to interpret when reading logs.